drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,660 @@
+ */
+int32_t adpd410x_reg_write(struct adpd410x_dev *dev, uint16_t address,
+			   uint16_t data)
+{
+	uint8_t buff[] = {0, 0, 0, 0};
+
+	if(dev->dev_type == ADPD4100) {
+		buff[0] = (address & 0x7f80) >> 7;
+		buff[1] = (address & 0x7f) << 1;
+		buff[1] &= 0xfe;
+		buff[2] = (data & 0xff00) >> 8;
+		buff[3] = data & 0xff;
+
+		return spi_write_and_read(dev->dev_ops.spi_phy_dev, buff, 4);
+	} else {
+		buff[0] = (address & 0x7f00) >> 8;
+		buff[0] |= 0x80;
+		buff[1] = address & 0xff;
+		buff[2] = (data & 0xff00) >> 8;
+		buff[3] = data & 0xff;
+
+		/* No stop bit */
+		return i2c_write(dev->dev_ops.i2c_phy_dev, buff, 4, 1);
+	}
+}
+
+/**
+ * @brief Do a software reset.
+ * @param dev - Device handler.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reset(struct adpd410x_dev *dev)
+{
+	int32_t ret;
+	uint16_t data;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_SYS_CTL, &data);
+	if(ret != SUCCESS)
+		return ret;
+	data |= BITM_SYS_CTL_SW_RESET;
+
+	return adpd410x_reg_write(dev, ADPD410X_REG_SYS_CTL, data);
+}
+
+/**
+ * @brief Set operation mode.
+ * @param dev - Device handler.
+ * @param mode - New operation mode.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_opmode(struct adpd410x_dev *dev,
+			    enum adpd410x_opmode mode)
+{
+	int32_t ret;
+	uint16_t data;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_OPMODE, &data);
+	if(ret != SUCCESS)
+		return ret;
+	data &= ~BITM_OPMODE_OP_MODE;
+	data |= mode;
+
+	return adpd410x_reg_write(dev, ADPD410X_REG_OPMODE, data);
+}
+
+/**
+ * @brief Set number of active time slots.
+ * @param dev - Device handler.
+ * @param timeslot_no - Last time slot to be enabled.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_last_timeslot(struct adpd410x_dev *dev,
+				   enum adpd410x_timeslots timeslot_no)
+{
+	int32_t ret;
+	uint16_t data;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_OPMODE, &data);
+	if(ret != SUCCESS)
+		return ret;
+	data &= ~BITM_OPMODE_TIMESLOT_EN;
+	data |= ((timeslot_no & 0x0f) << BITP_OPMODE_TIMESLOT_EN);
+
+	return adpd410x_reg_write(dev, ADPD410X_REG_OPMODE, data);
+}
+
+/**
+ * @brief Set device sampling frequency.
+ * @param dev - Device handler.
+ * @param sampling_freq - New sampling frequency.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_sampling_freq(struct adpd410x_dev *dev,
+				   uint32_t sampling_freq)
+{
+	int32_t ret;
+	uint32_t reg_load;
+	uint16_t reg_temp;
+
+	if((dev->clk_opt == ADPD410X_INTLFO_INTHFO) ||
+	    (dev->clk_opt == ADPD410X_INTLFO_EXTHFO)) {
+		ret = adpd410x_reg_read(dev, ADPD410X_REG_SYS_CTL, &reg_temp);
+		if(ret != SUCCESS)
+			return FAILURE;
+		if(reg_temp & BITP_SYS_CTL_LFOSC_SEL)
+			reg_load = ADPD410X_LOW_FREQ_OSCILLATOR_FREQ1;
+		else
+			reg_load = ADPD410X_LOW_FREQ_OSCILLATOR_FREQ2;
+	} else {
+		reg_load = dev->ext_lfo_freq;
+	}
+
+	reg_load /= sampling_freq;
```

Comment:
I'm not sure about this division here, remainder is not checked. Is this ok ?

Reply:
Yes, we do not use the remainder.